lib/client/api.go

```diff
@@ -2800,6 +2800,37 @@ func (tc *TeleportClient) getProxySSHPrincipal() string {
 	return proxyPrincipal
 }
 
+const unconfiguredPublicAddrMsg = `WARNING:
+
+The following error has occurred as Teleport does not recognise the address
+that is being used to connect to it. This usually indicates that the
+'public_addr' configuration option of the 'proxy_service' has not been
+set to match the address you are hosting the proxy on.
+
+If 'public_addr' is configured correctly, this could be an indicator of an
+attempted man-in-the-middle attack.
+`
+
+// formatConnectToProxyErr adds additional user actionable advice to errors
+// that are raised during ConnectToProxy.
+func formatConnectToProxyErr(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// Handles the error that occurs when you connect to the Proxy SSH service
+	// and the Proxy does not have a correct `public_addr` configured, and the
+	// system is configured with non-multiplexed ports.
+	if utils.IsHandshakeFailedError(err) {
+		principalStr := "not in the set of valid principals for given certificate"
+		if strings.Contains(trace.Unwrap(err).Error(), principalStr) {
+			return trace.WrapWithMessage(err, unconfiguredPublicAddrMsg)
```

Comment:
nit:

```suggestion
			return trace.Wrap(err, unconfiguredPublicAddrMsg)
```

Strangely, Wrap and WrapWithMessage do the same thing. I've been told to prefer the former, so passing on.